File: stream-chat-android/src/main/java/com/getstream/sdk/chat/adapter/ChannelListDiffCallback.java
package com.getstream.sdk.chat.adapter;

import java.util.List;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;
import io.getstream.chat.android.client.models.Channel;
import io.getstream.chat.android.client.models.Message;

import static com.getstream.sdk.chat.utils.LlcMigrationUtils.computeLastMessage;
import static com.getstream.sdk.chat.utils.LlcMigrationUtils.currentUserRead;
import static com.getstream.sdk.chat.utils.LlcMigrationUtils.equalsLastMessageDate;
import static com.getstream.sdk.chat.utils.LlcMigrationUtils.equalsName;
import static com.getstream.sdk.chat.utils.LlcMigrationUtils.equalsUserLists;
import static com.getstream.sdk.chat.utils.LlcMigrationUtils.getOtherUsers;
import static com.getstream.sdk.chat.utils.LlcMigrationUtils.lastMessagesAreTheSame;

public class ChannelListDiffCallback extends DiffUtil.Callback {
    private List<Channel> oldList, newList;

    public ChannelListDiffCallback(List<Channel> oldList, List<Channel> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList == null ? 0 : oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList == null ? 0 : newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        Channel oldItem = oldList.get(oldItemPosition);
        Channel newItem = newList.get(newItemPosition);
        return oldItem.getCid().equals(newItem.getCid());
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        Channel oldChannel = oldList.get(oldItemPosition);
        Channel newChannel = newList.get(newItemPosition);

        boolean contentTheSame = true;

        if (!oldChannel.getCid().equals(newChannel.getCid())) {
            contentTheSame = false;
        } else if (oldChannel.getUpdatedAt() == null && newChannel.getUpdatedAt() != null) {
            contentTheSame = false;
        } else if (newChannel.getUpdatedAt() != null && oldChannel.getUpdatedAt().getTime() < newChannel.getUpdatedAt().getTime()) {
            contentTheSame = false;
        } else if (!oldChannel.getExtraData().equals(newChannel.getExtraData())) {
            contentTheSame = false;
        } else if (!lastMessagesAreTheSame(oldChannel, newChannel)) {
            contentTheSame = false;
        } else if (currentUserRead(oldChannel, newChannel)) {
            contentTheSame = false;
        }

        return contentTheSame;
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {

        ChannelItemPayloadDiff diff = new ChannelItemPayloadDiff();

        Channel oldChannel = oldList.get(oldItemPosition);
        Channel newChannel = newList.get(newItemPosition);

        Message oldLastMessage = computeLastMessage(oldChannel);
        Message newLastMessage = computeLastMessage(newChannel);

        if (oldLastMessage != null && newLastMessage != null) {
            diff.lastMessage = !oldLastMessage.getId().equals(newLastMessage.getId());
        }

        diff.name = !equalsName(newChannel, oldChannel);
        diff.avatarView = !equalsUserLists(getOtherUsers(oldChannel.getMembers()), getOtherUsers(newChannel.getMembers()));
        diff.readState = currentUserRead(oldChannel, newChannel);
        diff.lastMessageDate = !equalsLastMessageDate(oldChannel, newChannel);

        return diff;
    }


}

File: stream-chat-android/src/main/java/com/getstream/sdk/chat/adapter/ChannelListDiffCallback.kt
package com.getstream.sdk.chat.adapter

import androidx.recyclerview.widget.DiffUtil
import com.getstream.sdk.chat.utils.extensions.computeLastMessage
import com.getstream.sdk.chat.utils.extensions.getOtherUsers
import io.getstream.chat.android.client.models.Channel
import io.getstream.chat.android.client.models.ChannelUserRead
import io.getstream.chat.android.client.models.User
import io.getstream.chat.android.client.models.name
import io.getstream.chat.android.livedata.ChatDomain

internal class ChannelListDiffCallback @JvmOverloads constructor(
    private val oldList: List<Channel>,
    private val newList: List<Channel>,
    private val currentUser: User = ChatDomain.instance().currentUser,
) : DiffUtil.Callback() {

    override fun getOldListSize(): Int = oldList.size

    override fun getNewListSize(): Int = newList.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldList[oldItemPosition].cid == newList[newItemPosition].cid
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldChannel = oldList[oldItemPosition]
        val newChannel = newList[newItemPosition]

        if ((oldChannel.cid != newChannel.cid) ||
            (oldChannel.updatedAt == null && newChannel.updatedAt != null) ||
            (newChannel.updatedAt != null && oldChannel.updatedAt!!.time < newChannel.updatedAt!!.time) ||
            (oldChannel.extraData != newChannel.extraData) ||
            !lastMessagesAreTheSame(oldChannel, newChannel) ||
            channelUserReadIsDifferent(oldChannel, newChannel) ||
            unreadCountIsDifferent(oldChannel, newChannel)
        ) {
            return false
        }
        return true
    }

    override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
        val oldChannel = oldList[oldItemPosition]
        val newChannel = newList[newItemPosition]
        return ChannelItemPayloadDiff(
            lastMessage = !lastMessagesAreTheSame(oldChannel, newChannel),
            name = !channelNameIsTheSame(newChannel, oldChannel),
            avatarView = !channelUsersAreTheSame(newChannel, oldChannel),
            readState = channelUserReadIsDifferent(oldChannel, newChannel),
            lastMessageDate = !channelLastMessageDatesAreTheSame(oldChannel, newChannel),
            unreadCount = unreadCountIsDifferent(oldChannel, newChannel),
        )
    }

    private fun lastMessagesAreTheSame(oldChannel: Channel, newChannel: Channel): Boolean {
        return oldChannel.computeLastMessage() == newChannel.computeLastMessage()
    }

    private fun channelNameIsTheSame(oldChannel: Channel, newChannel: Channel): Boolean {
        return oldChannel.name == newChannel.name
    }

    private fun channelUsersAreTheSame(oldChannel: Channel, newChannel: Channel): Boolean {
        val oldUserList = oldChannel.members.getOtherUsers(currentUser)
        val newUserList = newChannel.members.getOtherUsers(currentUser)

        if (oldUserList.size != newUserList.size) return false
        if (oldUserList.isEmpty() && newUserList.isEmpty()) return true

        for (i in oldUserList.indices) {
            if (oldUserList[i].id != newUserList[i].id) {
                return false
            }
        }
        return true
    }

    private fun channelLastMessageDatesAreTheSame(oldChannel: Channel, newChannel: Channel): Boolean {
        return oldChannel.lastMessageAt == newChannel.lastMessageAt
    }

    private fun unreadCountIsDifferent(oldChannel: Channel, newChannel: Channel): Boolean {
        return oldChannel.unreadCount != newChannel.unreadCount
    }

    private fun channelUserReadIsDifferent(oldChannel: Channel, newChannel: Channel): Boolean {
        val oldRead = getUserRead(oldChannel, currentUser.id)
        val newRead = getUserRead(newChannel, currentUser.id)
        return if (oldRead == null || newRead == null) {
            false
        } else {
            newRead.lastRead!!.after(oldRead.lastRead)
        }
    }

    private fun getUserRead(channel: Channel, userId: String): ChannelUserRead? {
        return channel.read.firstOrNull { it.getUserId() == userId }
    }
}
